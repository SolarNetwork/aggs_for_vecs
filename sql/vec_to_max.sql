-- smallint max lots
SELECT vec_to_max(smallints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- smallint max none
SELECT vec_to_max(smallints) FROM measurements WHERE sensor_id = -1;
-- smallint max one null
SELECT vec_to_max(smallints) FROM measurements WHERE sensor_id = 1;
-- smallint max array of nulls
SELECT vec_to_max(smallints) FROM measurements WHERE sensor_id = 2;
-- smallint max one not-null
SELECT vec_to_max(smallints) FROM measurements WHERE sensor_id = 4;
-- smallint max array of nulls and one other
SELECT vec_to_max(smallints) FROM measurements WHERE sensor_id IN (2, 4);

-- int max lots
SELECT vec_to_max(ints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- int max none
SELECT vec_to_max(ints) FROM measurements WHERE sensor_id = -1;
-- int max one null
SELECT vec_to_max(ints) FROM measurements WHERE sensor_id = 1;
-- int max array of nulls
SELECT vec_to_max(ints) FROM measurements WHERE sensor_id = 2;
-- int max one not-null
SELECT vec_to_max(ints) FROM measurements WHERE sensor_id = 4;
-- int max array of nulls and one other
SELECT vec_to_max(ints) FROM measurements WHERE sensor_id IN (2, 4);

-- bigint max lots
SELECT vec_to_max(bigints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- bigint max none
SELECT vec_to_max(bigints) FROM measurements WHERE sensor_id = -1;
-- bigint max one null
SELECT vec_to_max(bigints) FROM measurements WHERE sensor_id = 1;
-- bigint max array of nulls
SELECT vec_to_max(bigints) FROM measurements WHERE sensor_id = 2;
-- bigint max one not-null
SELECT vec_to_max(bigints) FROM measurements WHERE sensor_id = 4;
-- bigint max array of nulls and one other
SELECT vec_to_max(bigints) FROM measurements WHERE sensor_id IN (2, 4);

-- real max lots
SELECT vec_to_max(reals) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- real max none
SELECT vec_to_max(reals) FROM measurements WHERE sensor_id = -1;
-- real max one null
SELECT vec_to_max(reals) FROM measurements WHERE sensor_id = 1;
-- real max array of nulls
SELECT vec_to_max(reals) FROM measurements WHERE sensor_id = 2;
-- real max one not-null
SELECT vec_to_max(reals) FROM measurements WHERE sensor_id = 4;
-- real max array of nulls and one other
SELECT vec_to_max(reals) FROM measurements WHERE sensor_id IN (2, 4);

-- float8 max lots
SELECT vec_to_max(floats) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- float8 max none
SELECT vec_to_max(floats) FROM measurements WHERE sensor_id = -1;
-- float8 max one null
SELECT vec_to_max(floats) FROM measurements WHERE sensor_id = 1;
-- float8 max array of nulls
SELECT vec_to_max(floats) FROM measurements WHERE sensor_id = 2;
-- float8 max one not-null
SELECT vec_to_max(floats) FROM measurements WHERE sensor_id = 4;
-- float8 max array of nulls and one other
SELECT vec_to_max(floats) FROM measurements WHERE sensor_id IN (2, 4);

-- numeric max lots
SELECT vec_to_max(nums) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- numeric max none
SELECT vec_to_max(nums) FROM measurements WHERE sensor_id = -1;
-- numeric max one null
SELECT vec_to_max(nums) FROM measurements WHERE sensor_id = 1;
-- numeric max array of nulls
SELECT vec_to_max(nums) FROM measurements WHERE sensor_id = 2;
-- numeric max one not-null
SELECT vec_to_max(nums) FROM measurements WHERE sensor_id = 4;
-- numeric max array of nulls and one other
SELECT vec_to_max(nums) FROM measurements WHERE sensor_id IN (2, 4);
-- numeric max data 01
-- results expected to match:
-- WITH di AS (
--   SELECT idx_i, max(val_i) AS max_i
--   FROM measurements2 d, unnest(d.data_i) WITH ORDINALITY AS a(val_i, idx_i)
--   GROUP BY idx_i
-- )
-- SELECT array_agg(max_i ORDER BY idx_i) AS data_i_max FROM di
SELECT vec_to_max(data_i) FROM measurements2;
