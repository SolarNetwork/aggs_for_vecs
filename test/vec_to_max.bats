load test_helper

@test "int16 max" {
  result="$(query "SELECT vec_to_max(ints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4)")";
  echo $result;
  [ "$result" = "{1,2,3}" ]
}

# @test "int32 max" {
  # result="$(query "SELECT vec_to_max('{1,1,5,2,0}'::integer[])")";
  # [ "$result" = "0" ]
# }

# @test "int64 max" {
  # result="$(query "SELECT vec_to_max('{1,1,5,2,0}'::bigint[])")";
  # [ "$result" = "0" ]
# }

# @test "float4 max" {
  # result="$(query "SELECT vec_to_max('{1,1,5,2,0}'::real[])")";
  # [ "$result" = "0" ]
# }

@test "float8 max lots" {
  result="$(query "SELECT vec_to_max(floats) FROM measurements WHERE sensor_id IN (1, 2, 3, 4)")";
  echo $result;
  [ "$result" = "{1,2,3}" ]
}

@test "float8 max none" {
  result="$(query "SELECT vec_to_max(floats) FROM measurements WHERE sensor_id = -1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "float8 max one null" {
  result="$(query "SELECT vec_to_max(floats) FROM measurements WHERE sensor_id = 1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "float8 max array of nulls" {
  result="$(query "SELECT vec_to_max(floats) FROM measurements WHERE sensor_id = 2")";
  echo $result;
  [ "$result" = "{NULL,NULL,NULL}" ]
}

@test "float8 max one not-null" {
  result="$(query "SELECT vec_to_max(floats) FROM measurements WHERE sensor_id = 4")";
  echo $result;
  [ "$result" = "{1,NULL,2}" ]
}

@test "float8 max array of nulls and one other" {
  result="$(query "SELECT vec_to_max(floats) FROM measurements WHERE sensor_id IN (2, 4)")";
  echo $result;
  [ "$result" = "{1,NULL,2}" ]
}

@test "numeric max lots" {
  result="$(query "SELECT vec_to_max(nums) FROM measurements WHERE sensor_id IN (1, 2, 3, 4)")";
  echo $result;
  [ "$result" = "{1.23,2.34,3.45}" ]
}

@test "numeric max none" {
  result="$(query "SELECT vec_to_max(nums) FROM measurements WHERE sensor_id = -1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "numeric max one null" {
  result="$(query "SELECT vec_to_max(nums) FROM measurements WHERE sensor_id = 1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "numeric max array of nulls" {
  result="$(query "SELECT vec_to_max(nums) FROM measurements WHERE sensor_id = 2")";
  echo $result;
  [ "$result" = "{NULL,NULL,NULL}" ]
}

@test "numeric max one not-null" {
  result="$(query "SELECT vec_to_max(nums) FROM measurements WHERE sensor_id = 4")";
  echo $result;
  [ "$result" = "{1.23,NULL,2.34}" ]
}

@test "numeric max array of nulls and one other" {
  result="$(query "SELECT vec_to_max(nums) FROM measurements WHERE sensor_id IN (2, 4)")";
  echo $result;
  [ "$result" = "{1.23,NULL,2.34}" ]
}

# numeric max data 01 results calculated via:
# WITH di AS (
# 	SELECT idx_i, avg(val_i) AS avg_i
# 	FROM measurements2 d, unnest(d.data_i) WITH ORDINALITY AS a(val_i, idx_i)
# 	GROUP BY idx_i
# )
# SELECT array_agg(avg_i ORDER BY idx_i) AS data_i_avg FROM di

@test "numeric max data 01" {
  result="$(query "SELECT vec_to_max(data_i) FROM measurements2")";
  echo $result;
  [ "$result" = "{23200619,48587.514387,234.99962,50.5,13161880,1,332.20728,23730754,23623545}" ]
}
