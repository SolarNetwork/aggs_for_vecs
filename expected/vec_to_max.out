-- smallint max lots
SELECT vec_to_max(smallints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_max 
------------
 {1,2,3}
(1 row)

-- smallint max none
SELECT vec_to_max(smallints) FROM measurements WHERE sensor_id = -1;
 vec_to_max 
------------
 
(1 row)

-- smallint max one null
SELECT vec_to_max(smallints) FROM measurements WHERE sensor_id = 1;
 vec_to_max 
------------
 
(1 row)

-- smallint max array of nulls
SELECT vec_to_max(smallints) FROM measurements WHERE sensor_id = 2;
    vec_to_max    
------------------
 {NULL,NULL,NULL}
(1 row)

-- smallint max one not-null
SELECT vec_to_max(smallints) FROM measurements WHERE sensor_id = 4;
 vec_to_max 
------------
 {1,NULL,2}
(1 row)

-- smallint max array of nulls and one other
SELECT vec_to_max(smallints) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_max 
------------
 {1,NULL,2}
(1 row)

-- int max lots
SELECT vec_to_max(ints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_max 
------------
 {1,2,3}
(1 row)

-- int max none
SELECT vec_to_max(ints) FROM measurements WHERE sensor_id = -1;
 vec_to_max 
------------
 
(1 row)

-- int max one null
SELECT vec_to_max(ints) FROM measurements WHERE sensor_id = 1;
 vec_to_max 
------------
 
(1 row)

-- int max array of nulls
SELECT vec_to_max(ints) FROM measurements WHERE sensor_id = 2;
    vec_to_max    
------------------
 {NULL,NULL,NULL}
(1 row)

-- int max one not-null
SELECT vec_to_max(ints) FROM measurements WHERE sensor_id = 4;
 vec_to_max 
------------
 {1,NULL,2}
(1 row)

-- int max array of nulls and one other
SELECT vec_to_max(ints) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_max 
------------
 {1,NULL,2}
(1 row)

-- bigint max lots
SELECT vec_to_max(bigints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_max 
------------
 {1,2,3}
(1 row)

-- bigint max none
SELECT vec_to_max(bigints) FROM measurements WHERE sensor_id = -1;
 vec_to_max 
------------
 
(1 row)

-- bigint max one null
SELECT vec_to_max(bigints) FROM measurements WHERE sensor_id = 1;
 vec_to_max 
------------
 
(1 row)

-- bigint max array of nulls
SELECT vec_to_max(bigints) FROM measurements WHERE sensor_id = 2;
    vec_to_max    
------------------
 {NULL,NULL,NULL}
(1 row)

-- bigint max one not-null
SELECT vec_to_max(bigints) FROM measurements WHERE sensor_id = 4;
 vec_to_max 
------------
 {1,NULL,2}
(1 row)

-- bigint max array of nulls and one other
SELECT vec_to_max(bigints) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_max 
------------
 {1,NULL,2}
(1 row)

-- real max lots
SELECT vec_to_max(reals) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_max 
------------
 {1,2,3}
(1 row)

-- real max none
SELECT vec_to_max(reals) FROM measurements WHERE sensor_id = -1;
 vec_to_max 
------------
 
(1 row)

-- real max one null
SELECT vec_to_max(reals) FROM measurements WHERE sensor_id = 1;
 vec_to_max 
------------
 
(1 row)

-- real max array of nulls
SELECT vec_to_max(reals) FROM measurements WHERE sensor_id = 2;
    vec_to_max    
------------------
 {NULL,NULL,NULL}
(1 row)

-- real max one not-null
SELECT vec_to_max(reals) FROM measurements WHERE sensor_id = 4;
 vec_to_max 
------------
 {1,NULL,2}
(1 row)

-- real max array of nulls and one other
SELECT vec_to_max(reals) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_max 
------------
 {1,NULL,2}
(1 row)

-- float8 max lots
SELECT vec_to_max(floats) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_max 
------------
 {1,2,3}
(1 row)

-- float8 max none
SELECT vec_to_max(floats) FROM measurements WHERE sensor_id = -1;
 vec_to_max 
------------
 
(1 row)

-- float8 max one null
SELECT vec_to_max(floats) FROM measurements WHERE sensor_id = 1;
 vec_to_max 
------------
 
(1 row)

-- float8 max array of nulls
SELECT vec_to_max(floats) FROM measurements WHERE sensor_id = 2;
    vec_to_max    
------------------
 {NULL,NULL,NULL}
(1 row)

-- float8 max one not-null
SELECT vec_to_max(floats) FROM measurements WHERE sensor_id = 4;
 vec_to_max 
------------
 {1,NULL,2}
(1 row)

-- float8 max array of nulls and one other
SELECT vec_to_max(floats) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_max 
------------
 {1,NULL,2}
(1 row)

-- numeric max lots
SELECT vec_to_max(nums) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
    vec_to_max    
------------------
 {1.23,2.34,3.45}
(1 row)

-- numeric max none
SELECT vec_to_max(nums) FROM measurements WHERE sensor_id = -1;
 vec_to_max 
------------
 
(1 row)

-- numeric max one null
SELECT vec_to_max(nums) FROM measurements WHERE sensor_id = 1;
 vec_to_max 
------------
 
(1 row)

-- numeric max array of nulls
SELECT vec_to_max(nums) FROM measurements WHERE sensor_id = 2;
    vec_to_max    
------------------
 {NULL,NULL,NULL}
(1 row)

-- numeric max one not-null
SELECT vec_to_max(nums) FROM measurements WHERE sensor_id = 4;
    vec_to_max    
------------------
 {1.23,NULL,2.34}
(1 row)

-- numeric max array of nulls and one other
SELECT vec_to_max(nums) FROM measurements WHERE sensor_id IN (2, 4);
    vec_to_max    
------------------
 {1.23,NULL,2.34}
(1 row)

-- numeric max data 01
-- results expected to match:
-- WITH di AS (
--   SELECT idx_i, max(val_i) AS max_i
--   FROM measurements2 d, unnest(d.data_i) WITH ORDINALITY AS a(val_i, idx_i)
--   GROUP BY idx_i
-- )
-- SELECT array_agg(max_i ORDER BY idx_i) AS data_i_max FROM di
SELECT vec_to_max(data_i) FROM measurements2;
                                  vec_to_max                                   
-------------------------------------------------------------------------------
 {23200619,48587.514387,234.99962,50.5,13161880,1,332.20728,23730754,23623545}
(1 row)

